frametrim/ft_main.cpp

```diff
@@ -0,0 +1,278 @@
+                      << " call:" << call->no;
+
+        call.reset(p.parse_call());
+        ++calls_in_this_frame;
+    }
+    trimmer.finalize();
+    std::cerr << "\nDone scanning frames\n";
+
+    trace::Writer writer;
+    if (!writer.open(out_filename.c_str(), p.getVersion(), p.getProperties())) {
+        std::cerr << "error: failed to create " << out_filename << "\n";
+        return 2;
+    }
+
+    auto call_ids = trimmer.getSortedCallIds();
+    std::cerr << "Write output file\n";
+
+    p.close();
+    p.open(filename);
+    call.reset(p.parse_call());
+
+    auto callid_itr = call_ids.begin();
+
+    std::cerr << "Copying " << call_ids.size() << " calls\n";
+
+    while (call && callid_itr != call_ids.end()) {
+        while (call->no != *callid_itr)
+            call.reset(p.parse_call());
+
+        writer.writeCall(call.get());
+        call.reset(p.parse_call());
+        ++callid_itr;
+    }
+
+    if (options.top_frame_call_counts) {
+        unsigned count = options.top_frame_call_counts;
+        std::cerr << "Calls per frame:\n";
+        while (count-- && !calls_in_frame.empty()) {
+            auto fc = calls_in_frame.top();
+            calls_in_frame.pop();
+            std::cerr << "  Frame[" << fc.second << "] = " << fc.first << "\n";
+        }
+    }
+
+    return 0;
+}
+
+
+int main(int argc, char **argv)
+{
+    struct trim_options options;
+    options.frames = trace::CallSet(trace::FREQUENCY_NONE);
+    options.top_frame_call_counts = false;
+    options.keep_all_states = false;
+
+    int opt;
+    while ((opt = getopt_long(argc, argv, shortOptions, longOptions, nullptr)) != -1) {
+        switch (opt) {
+        case 'h':
+            usage();
+            return 0;
+        case FRAMES_OPT:
+            options.frames.merge(optarg);
+            break;
+        case SETUPFRAMES_OPT:
+            options.setupframes.merge(optarg);
+            break;
+        case 'o':
+            options.output = optarg;
+            break;
+        case 't':
+            options.top_frame_call_counts = atoi(optarg);
+            break;
+        case 'k':
+            options.keep_all_states = true;
+            break;
+        default:
+            std::cerr << "error: unexpected option `" << (char)opt << "`\n";
+            usage();
+            return 1;
+        }
+    }
+
+    /*if (options.frames.getFirst() != options.frames.getLast())  {
+        std::cerr << "error: Must give exactly one frame\n";
+        return 1;
+    }*/
+
+    if (optind >= argc) {
+        std::cerr << "error: apitrace trim requires a trace file as an argument.\n";
+        usage();
+        return 1;
+    }
+
+    if (argc > optind + 1) {
+        std::cerr << "error: extraneous arguments:";
+        for (int i = optind + 1; i < argc; i++) {
+            std::cerr << " " << argv[i];
+        }
+        std::cerr << "\n";
+        usage();
+        return 1;
+    }
+
+    return trim_to_frame(argv[optind], options);
```

Comment:
This utility should have a warning stating there are some limitations, perhaps pointing to `frametrim.markdown`

Reply:
I've added a message when the program is run that points out that one should check the result and hinted to the markdown file for further details.